src/NHibernate/Linq/Visitors/SelectClauseNominator.cs

```diff
@@ -114,11 +114,14 @@ private bool IsRegisteredFunction(Expression expression)
 			if (expression.NodeType == ExpressionType.Call)
 			{
 				var methodCallExpression = (MethodCallExpression) expression;
-				IHqlGeneratorForMethod methodGenerator;
-				if (_functionRegistry.TryGetGenerator(methodCallExpression.Method, out methodGenerator))
+				if (_functionRegistry.TryGetGenerator(methodCallExpression.Method, out var methodGenerator))
```

Comment:
New lookup of `IRuntimeMethodHqlGenerator`...

Here it is done only for methods inside the select clause. This part is responsible for extracting things deemed not translatable to SQL out of the query, for running them with .Net Runtime on the results.
It considers that methods belonging to instances not part of the query cannot be translated. In general this is right, unless the translation does not care for the instance state, which is the case of the `Random` generator. So here we need to ask any registered generator for the method if it cares about the instance.